feat: Docker Image and GCS Integration

realjockel opened the pull request:
## Summary

  - Added Docker containerization support for easy deployment
  - Integrated Google Cloud Storage (GCS) for persistent state management across runs
  - Enhanced CLI with GCS configuration options and chain-based partitioning

##  Changes

 ###  Docker Support

  - `Dockerfile`: Production-ready multi-stage build with Node 18 slim base image
    - Non-root user execution for security
    - Proper directory permissions for store volume
  - `.dockerignore`: Excludes development files and local store from build context

 ###  GCS Integration

  - New GCS sync module (src/infrastructure/storage/gcs-sync.ts):
    - Downloads existing store data from GCS before processing
    - Uploads updated store data after pipeline completion
    - Supports chain-based path partitioning for multi-chain deployments
    - Handles partial uploads on pipeline failures
  - CLI enhancements:
    - `--gcs-bucket`: Specify GCS bucket for persistent storage
    - `--gcs-project:` Optional GCP project ID (uses default credentials if not specified)
    - `--chain`: Chain identifier for GCS path partitioning (e.g., aep-fee-tracker/{chain}/)
    
  - Orchestrator updates:
    - Integrated GCS sync into the main pipeline workflow
    - Downloads state before processing to continue from previous runs
    - Uploads results after successful completion
    - Attempts partial upload even on pipeline failures

  Usage Examples
```
  # Run with GCS persistence
  aep --rpc-url https://rpc.example.com \
      --gcs-bucket my-bucket \
      --chain logx
```
```
  # Docker deployment with GCS
  docker build -t aep-fee-tracker .
  docker run aep-fee-tracker \
      --rpc-url $RPC_URL \
      --gcs-bucket $GCS_BUCKET \
      --chain $CHAIN
```

realjockel committed:
feat: Docker Image and GCS Integration

yahgwai commented on src/cli/parse-arguments.ts:
```diff
@@ -6,20 +6,34 @@ export interface ParsedArguments {
+  --store-dir <path>      Directory for storing data
+  --gcs-bucket <bucket>   GCS bucket name for persistent storage
+  --gcs-project <project> GCP project ID (optional, uses default credentials)
+  --chain <chain>         Chain identifier for GCS path partitioning`;
```
I suppose we could dynamically fetch a chain id from the provider rather than specifying here - but if we want something more readable this seems like a good approach

realjockel replied:
yeah that's what i thought. I wanted to partition the gcs bucket by chain={chain-name} for readability

yahgwai commented on src/infrastructure/orchestration/orchestrator.ts:
```diff
@@ -16,15 +17,50 @@ export async function orchestrate(config: Configuration): Promise<void> {
+    // Upload updated store data to GCS if configured
+    if (gcsSync) {
+      console.log("📤 Uploading updated store data to GCS...");
+      await gcsSync.uploadStore(config.storeDirectory);
```
errors from here will hit the catch block, and try again. that could ok I guess? if we dont need that we could put this sync in a `finally` block, then it will trigger on error or success

yahgwai commented on src/cli/parse-arguments.ts:
```diff
@@ -6,20 +6,34 @@ export interface ParsedArguments {
+  --start-date <date>     Start date in YYYY-MM-DD format
+  --end-date <date>       End date in YYYY-MM-DD format
+  --store-dir <path>      Directory for storing data
+  --gcs-bucket <bucket>   GCS bucket name for persistent storage
```
should we validate somewhere that either all these options are provided or non?

yahgwai commented on src/cli/configuration.ts:
```diff
@@ -30,5 +30,17 @@ export function createConfiguration(
+  }
+
+  if (parsedArgs["gcs-project"]) {
+    config.gcsProject = parsedArgs["gcs-project"];
```
what's this command line arg used for? i dont see usage

realjockel replied:
actually, i wanted to provide this to the storageClient. But it is unnecessary, because we are using application default credentials in GCP and don't need to explicitly provide the gcp project. Gonna remove it.

yahgwai commented on src/infrastructure/storage/gcs-sync.ts:
```diff
@@ -0,0 +1,176 @@
+  constructor(
+    bucketName: string,
+    remotePath: string = "aep-fee-tracker",
+    chain?: string,
```
Personal preference, but I usually opt for non optional params except in certain situations. in this case i dont think you need the default remotePath either since you're supplying that in the one place we use this

yahgwai commented on src/infrastructure/storage/gcs-sync.ts:
```diff
@@ -0,0 +1,176 @@
+        }
+
+        // Download file
+        await file.download({ destination: localFilePath });
```
Does this overwrite existing files or keep local? dunno which behaviour your want

realjockel replied:
overwrite existing local files. Source of truth for runs with GCS enabled should be the remote bucket

realjockel replied:
or do you have a different opinion?

yahgwai replied:
I think that's fine. One edge case that I can see might be where the state gets corrupted, or you want to overwrite it eg because we change the format or something. It might be nice to do a local run then push the result to gcs. But I dont think you need to build support for that, that can be done more ad hoc.

yahgwai commented on src/infrastructure/storage/gcs-sync.ts:
```diff
@@ -0,0 +1,176 @@
+          metadata: {
+            contentType: fullPath.endsWith(".json")
+              ? "application/json"
+              : "application/octet-stream",
```
Interesting, do we expect non json here? otherwise I'd be tempted to error out if we the file doesnt end with .json, rather than silently continuing

realjockel replied:
agree, we should only expect json here.

yahgwai commented on src/infrastructure/storage/gcs-sync.ts:
```diff
@@ -0,0 +1,176 @@
+  /**
+   * Check if store exists in GCS
+   */
+  async storeExists(): Promise<boolean> {
```
This doesnt appear to be used

realjockel replied:
correct, it's unnecessary. the downloadStore is handling each case gracefully, so no need for this.

yahgwai commented on src/infrastructure/storage/gcs-sync.ts:
```diff
@@ -0,0 +1,176 @@
+  /**
+   * Get store info for debugging
+   */
+  async getStoreInfo(): Promise<{ fileCount: number; totalSize: number }> {
```
also not used?

yahgwai committed:
chore: removing logging from tests

realjockel committed:
fix: address PR review comments for GCS integration

yahgw-ai committed:
fix: align CLI argument help text spacing

Standardized spacing in argument descriptions to ensure consistent formatting
between implementation and tests. This fixes the failing test that expected
different spacing in the help text output.
